django/db/models/fields/generated.py

```diff
@@ -0,0 +1,159 @@
+from django.core import checks
+from django.db import connections
+from django.db.models.sql import Query
+
+from . import Field
+
+__all__ = ["GeneratedField"]
+
+
+class GeneratedField(Field):
+    generated = True
+    db_returning = True
+
+    _query = None
+    _resolved_expression = None
+    output_field = None
+
+    def __init__(self, *, expression, output_field=None, db_persist=True, **kwargs):
+        self.expression = expression
+        self._output_field = output_field
+        self.db_persist = db_persist
+
+        assert "editable" not in kwargs, "Cannot set `editable` for GeneratedField"
+        assert "blank" not in kwargs, "Cannot set `blank` for GeneratedField"
+        assert "default" not in kwargs, "Cannot set `default` for GeneratedField"
```

Comment:
This is a user-facing error so we shouldn't use `assert` (check out ticket-32508).

Reply:
Yeah, I was wondering whether to fix that myself or wait for the parent PR. Do you have an opinion about the best error type?

Reply:
> Yeah, I was wondering whether to fix that myself or wait for the parent PR.

Leaving comments in two PRs is hard to track, so I closed the source PR.

> Do you have an opinion about the best error type?

Maybe `ValueError`, e.g.
```python
if kwargs.get("editable", False) is True:
    raise ValueError("GeneratedField cannot be editable.")
```

Reply:
> > Yeah, I was wondering whether to fix that myself or wait for the parent PR.
> 
> Leaving comments in two PRs is hard to track, so I closed the source PR.

I think the correct title for this PR is now "Add support for function-based virtual fields" because is not Oracle-specific anymore, and more people would review it.